cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -609,34 +707,44 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
                          std::move(aggregate_schema));
     }
 
-    case substrait::Rel::RelTypeCase::kExtensionLeaf: {
-      const auto& ext = rel.extension_leaf();
-      ARROW_ASSIGN_OR_RAISE(
-          auto ext_leaf_decl,
-          conversion_options.extension_provider->MakeRel({}, ext.detail(), ext_set));
-      return ProcessEmit(ext, std::move(ext_leaf_decl), ext_leaf_decl.output_schema);
-    }
-    case substrait::Rel::RelTypeCase::kExtensionSingle: {
-      const auto& ext = rel.extension_single();
-      ARROW_ASSIGN_OR_RAISE(DeclarationInfo input,
-                            FromProto(ext.input(), ext_set, conversion_options));
-      ARROW_ASSIGN_OR_RAISE(
-          auto ext_single_decl,
-          conversion_options.extension_provider->MakeRel({input}, ext.detail(), ext_set));
-      return ProcessEmit(ext, std::move(ext_single_decl), ext_single_decl.output_schema);
-    }
+    case substrait::Rel::RelTypeCase::kExtensionLeaf:
+    case substrait::Rel::RelTypeCase::kExtensionSingle:
     case substrait::Rel::RelTypeCase::kExtensionMulti: {
-      const auto& ext = rel.extension_multi();
-      std::vector<DeclarationInfo> inputs;
-      for (const auto& input : ext.inputs()) {
-        ARROW_ASSIGN_OR_RAISE(auto input_info,
-                              FromProto(input, ext_set, conversion_options));
-        inputs.push_back(std::move(input_info));
-      }
+      std::vector<DeclarationInfo> ext_rel_inputs;
       ARROW_ASSIGN_OR_RAISE(
-          auto ext_multi_decl,
-          conversion_options.extension_provider->MakeRel(inputs, ext.detail(), ext_set));
-      return ProcessEmit(ext, std::move(ext_multi_decl), ext_multi_decl.output_schema);
+          auto ext_rel_info,
+          GetExtensionRelationInfo(rel, ext_set, conversion_options, &ext_rel_inputs));
+      const auto& ext_decl_info = ext_rel_info.decl_info;
+      auto ext_common_opt = GetExtensionRelCommon(rel);
+      bool has_emit = ext_common_opt && ext_common_opt->emit_kind_case() ==
```

Comment:
Does emit kind get set somewhere? I don't see it from the substrait plan produced by ibis:
https://github.com/ibis-project/ibis-substrait/issues/423#issuecomment-1314065117

Reply:
In the protobuf it is defined as:

```
  oneof emit_kind {
    // The underlying relation is output as is (no reordering or projection of columns)
    Direct direct = 1;
    // Allows to control for order and inclusion of fields
    Emit emit = 2;
  }
```

From Ibis we see:

```
common {
  emit {
    output_mapping: 6
    output_mapping: 7
    output_mapping: 8
    output_mapping: 9
  }
}
```

So in this case `emit_kind` will be set to `emit`.